[lldb][NFC] Change ObjectFile argument type

jasonmolenda opened the pull request:
The ObjectFile plugin interface accepts an optional DataBufferSP argument.  If the caller has the contents of the binary, it can provide this in that DataBufferSP. The ObjectFile subclasses in their CreateInstance methods will fill in the DataBufferSP with the actual binary contents if it is not set.
ObjectFile base class creates an ivar DataExtractor from the DataBufferSP passed in.

My next patch will be a caller that creates a VirtualDataExtractor with the binary data, and needs to pass that in to the ObjectFile plugin, instead of the bag-of-bytes DataBufferSP.  It builds on the previous patch changing ObjectFile's ivar from DataExtractor to DataExtractorSP so I could pass in a subclass in the shared ptr. And it will be using the VirtualDataExtractor that Jonas added in https://github.com/llvm/llvm-project/pull/168802

No behavior is changed by the patch; we're simply moving the creation of the DataExtractor to the caller, instead of a DataBuffer that is immediately used to set up the ObjectFile DataExtractor.  The patch is a bit complicated because all of the ObjectFile subclasses have to initialize their DataExtractor to pass in to the base class.

I ran the testsuite on macOS and on AArch64 Ubutnu.  (btw David, I ran it under qemu on my M4 mac with SME-no-SVE again, Ubuntu 25.10, checked lshw(1) cpu capabilities, and qemu doesn't seem to be virtualizing the SME, that explains why the testsuite passes)

rdar://148939795

jasonmolenda committed:
[lldb][NFC] Change ObjectFile argument type

The ObjectFile plugin interface accepts an optional
DataBufferSP argument.  If the caller has the contents
of the binary, it can provide this in that DataBufferSP.
The ObjectFile subclasses in their CreateInstance methods
will fill in the DataBufferSP with the actual binary contents
if it is not set.
ObjectFile base class creates an ivar DataExtractor from the
DataBufferSP passed in.

My next patch will be a caller that creates a VirtualDataExtractor
with the binary data, and needs to pass that in to the ObjectFile
plugin, instead of the bag-of-bytes DataBufferSP.  It builds on
the previous patch changing ObjectFile's ivar from DataExtractor
to DataExtractorSP so I could pass in a subclass in the shared ptr.
And it will be using the VirtualDataExtractor that Jonas added in
https://github.com/llvm/llvm-project/pull/168802

No behavior is changed by the patch; we're simply moving the creation
of the DataExtractor to the caller, instead of a DataBuffer that
is immediately used to set up the ObjectFile DataExtractor.  The
patch is a bit complicated because all of the ObjectFile subclasses
have to initialize their DataExtractor to pass in to the base class.

I ran the testsuite on macOS and on AArch64 Ubutnu.  (btw David, I
ran it under qemu on my M4 mac with SME-no-SVE again, Ubuntu 25.10,
checked lshw(1) cpu capabilities, and qemu doesn't seem to be
virtualizing the SME, that explains why the testsuite passes)

rdar://148939795

DavidSpickett reviewed (commented):
Since @JDevlieghere has more context for this effort, they can be the approver.

jasonmolenda committed:
Fix DataExtractor.h method descriptions based on
David's feedback.

JDevlieghere commented on lldb/source/Plugins/ObjectContainer/BSD-Archive/ObjectContainerBSDArchive.h:
```diff
@@ -166,9 +169,9 @@ class ObjectContainerBSDArchive : public lldb_private::ObjectContainer {
-                                        ///gets modified
+    lldb::DataExtractorSP m_data_sp; ///< The data for this object container
+                                     /// so we don't lose data if the .a files
+                                     /// gets modified
```
```suggestion
    /// The data for this object container so we don't lose data if the .a files
    /// gets modified.
    lldb::DataExtractorSP m_data_sp;
```

JDevlieghere commented on lldb/source/Symbol/ObjectFile.cpp:
```diff
@@ -260,8 +268,13 @@ ObjectFile::ObjectFile(const lldb::ModuleSP &module_sp,
-    m_data_nsp->SetData(data_sp, data_offset, length);
+  if (extractor_sp && extractor_sp->GetByteSize() > 0) {
+    m_data_nsp = extractor_sp;
+    // the offset & length fields may be specifying a subset of the
```
```suggestion
    // The offset & length fields may be specifying a subset of the
```

JDevlieghere reviewed (commented):
Motivation for this patch aside, it seems equally reasonable to pass around and store a `DataExtractorSP` as it is to use a `DataBufferSP`. The patch looks mostly mechanical (though I know there were a few tricky edge cases). 

General observations:

1. It would be nice if we could consistently use `extractor_sp` for a `DataExtractorSP` and `data_sp` for a `DataBufferSP`. There's a few cases where you changed the type but not the member, presumably to limit the number of lines changed. I think it would be worth the churn to avoid confusion/surprises down the line. 
2. I'm not a huge fan of checking the `ByteSize`. I would prefer an `operator bool` opr a `HasData()` that checks that instead.

jasonmolenda committed:
Update lldb/source/Plugins/ObjectContainer/BSD-Archive/ObjectContainerBSDArchive.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Update lldb/source/Symbol/ObjectFile.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Update lldb/source/Utility/DataExtractor.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda commented on lldb/source/Symbol/ObjectFile.cpp:
```diff
@@ -66,14 +68,14 @@ ObjectFile::FindPlugin(const lldb::ModuleSP &module_sp, const FileSpec *file,
     if (file_exists && module_sp->GetObjectName()) {
       ObjectFileSP object_file_sp = CreateObjectFromContainer(
-          module_sp, file, file_offset, file_size, data_sp, data_offset);
+          module_sp, file, file_offset, file_size, DataBufferSP(), data_offset);
```
This CreateObjectFromContainer takes a DataBufferSP().  This block opens with `!extractor_sp || extractor_sp->GetByteSize() == 0` so it might be an empty DataExtractorSP; we can't pass `extractor_sp->GetSharedDataBuffer()` here.

jasonmolenda commented on lldb/source/Plugins/ObjectFile/wasm/ObjectFileWasm.cpp:
```diff
@@ -139,44 +139,48 @@ void ObjectFileWasm::Terminate() {
 
-  if (!data_sp) {
-    data_sp = MapFileData(*file, length, file_offset);
+  if (!extractor_sp || extractor_sp->GetByteSize() == 0) {
```
yeah, I'm not sure I like using an operator bool here, a `DataExtractor::HasData()` method is the approach I would like.  But you're right the idiom is used multiple times in this PR.

jasonmolenda committed:
Add a DataExtractor::HasData() method, update
callers of GetByteSize()==0 to use it.

jasonmolenda committed:
Implement stricter naming convention of
DataExtractos being "extractor", DataBuffers being "data"
in all the files being touched by this PR.

DavidSpickett commented:
> I didn't want to get too deep into details not directly related to this PR, but the overall project is a mac specific one.

Thanks for explaining and now you've explained, I feel justified in getting Jonas to review :)

JDevlieghere reviewed (approved):
LGTM
